File: sdks/typescript/src/pagination.ts
import type { StarlingClient } from "./client.js";
import type { Starling } from "./generated/index.js";
import type { BaseRequestOptions } from "./generated/BaseClient.js";
import { encodePathParam } from "./generated/core/url/index.js";

export class PaginationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "PaginationError";
  }
}

export interface FeedIteratorOptions extends Omit<BaseRequestOptions, "queryParams" | "additionalBodyParameters" | "stream"> {
  /** Stop after this many pages. Omit to consume the complete range. */
  maxPages?: number;
}

const parameters = new Set(["minTransactionTimestamp", "maxTransactionTimestamp", "cursor", "pageToFetch"]);

/** Lazily follow the paginated feed's next links, without changing accounts or date range. */
export async function* iterateFeed(client: StarlingClient, request: Starling.ListPageFeedRequest, options: FeedIteratorOptions = {}): AsyncGenerator<Starling.FeedItem> {
  const { maxPages, ...requestOptions } = options;
  if ("queryParams" in options && options.queryParams) throw new TypeError("Pass feed filters directly; queryParams is unsupported by iterateFeed");
  if (maxPages !== undefined && (!Number.isSafeInteger(maxPages) || maxPages < 1)) throw new RangeError("maxPages must be a positive integer");
  const min = Date.parse(request.minTransactionTimestamp);
  const max = Date.parse(request.maxTransactionTimestamp);
  if (!Number.isFinite(min) || !Number.isFinite(max) || min > max) throw new RangeError("Provide a valid minimum and maximum transaction timestamp");
  const base = new URL(`/api/v2/feed/account/${encodePathParam(request.accountUid)}/category/${encodePathParam(request.categoryUid)}/paginated-transactions`, client.baseUrl);
  const seen = new Set<string>();
  let current = { ...request };
  for (let count = 0; ; count++) {
    options.abortSignal?.throwIfAborted();
    const marker = JSON.stringify([current.cursor ?? null, current.cursor ? (current.pageToFetch ?? "NEXT") : null]);
    if (seen.has(marker)) throw new PaginationError("Starling returned a repeated pagination cursor");
    seen.add(marker);
    const page = await client.feed.listPage(current, requestOptions);
    if (!Array.isArray(page?.feedItems) || !page.links) throw new PaginationError("Starling returned an invalid feed page");
    for (const item of page.feedItems) {
      options.abortSignal?.throwIfAborted();
      yield item;
    }
    if (!page.links.next || (maxPages !== undefined && count + 1 >= maxPages)) return;
    let next: URL;
    try { next = new URL(page.links.next, base); }
    catch { throw new PaginationError("Starling returned an invalid next link"); }
    if (next.origin !== base.origin || next.pathname !== base.pathname || next.username || next.password || next.hash) throw new PaginationError("The next link changed the API origin or feed endpoint");
    for (const key of next.searchParams.keys()) {
      if (!parameters.has(key) || next.searchParams.getAll(key).length !== 1) throw new PaginationError("The next link contains unsupported or duplicate parameters");
    }
    const minimum = next.searchParams.get("minTransactionTimestamp");
    const maximum = next.searchParams.get("maxTransactionTimestamp");
    if (!minimum || !maximum || Date.parse(minimum) !== min || Date.parse(maximum) !== max) throw new PaginationError("The next link changed the requested date range");
    const cursor = next.searchParams.get("cursor");
    const direction = next.searchParams.get("pageToFetch") ?? "NEXT";
    if (!cursor || direction !== "NEXT") throw new PaginationError("The next link has no forward pagination cursor");
    current = { ...request, minTransactionTimestamp: minimum, maxTransactionTimestamp: maximum, cursor, pageToFetch: "NEXT" };
  }
}

File: sdks/python/src/starling_bank/pagination.py
"""Lazy iteration over Starling's paginated transaction feed."""

from __future__ import annotations

from datetime import datetime
from typing import AsyncIterator, Iterator
from urllib.parse import parse_qs, quote, urljoin, urlsplit

from .client import AsyncStarlingClient, StarlingClient
from .generated.core.request_options import RequestOptions
from .generated.types.feed_item import FeedItem


class PaginationError(Exception):
    pass


class _FeedPagination:
    def __init__(self, base_url, account_uid, category_uid, minimum, maximum, max_pages):
        if minimum.tzinfo is None or maximum.tzinfo is None or minimum > maximum:
            raise ValueError("Provide timezone-aware minimum and maximum transaction timestamps in order")
        if max_pages is not None and (not isinstance(max_pages, int) or isinstance(max_pages, bool) or max_pages < 1):
            raise ValueError("max_pages must be a positive integer")
        path = f"/api/v2/feed/account/{quote(account_uid, safe='')}/category/{quote(category_uid, safe='')}/paginated-transactions"
        self.base = urlsplit(base_url + path)
        self.minimum, self.maximum = minimum, maximum
        self.request = dict(account_uid=account_uid, category_uid=category_uid, min_transaction_timestamp=minimum, max_transaction_timestamp=maximum)
        self.seen: set[tuple[str | None, str | None]] = set()

    def take(self):
        marker = (self.request.get("cursor"), self.request.get("page_to_fetch"))
        if marker in self.seen:
            raise PaginationError("Starling returned a repeated pagination cursor")
        self.seen.add(marker)
        return self.request

    def advance(self, link):
        try:
            next_url = urlsplit(urljoin(self.base.geturl(), link))
            if (next_url.scheme, next_url.hostname, next_url.port, next_url.path) != (self.base.scheme, self.base.hostname, self.base.port, self.base.path) or next_url.username or next_url.password or next_url.fragment:
                raise PaginationError("The next link changed the API origin or feed endpoint")
            values = parse_qs(next_url.query, keep_blank_values=True)
            allowed = {"minTransactionTimestamp", "maxTransactionTimestamp", "cursor", "pageToFetch"}
            if any(key not in allowed or len(value) != 1 for key, value in values.items()):
                raise PaginationError("The next link contains unsupported or duplicate parameters")
            minimum = datetime.fromisoformat(values["minTransactionTimestamp"][0])
            maximum = datetime.fromisoformat(values["maxTransactionTimestamp"][0])
            if minimum != self.minimum or maximum != self.maximum:
                raise PaginationError("The next link changed the requested date range")
            cursor = values["cursor"][0]
            if not cursor or values.get("pageToFetch", ["NEXT"])[0] != "NEXT":
                raise PaginationError("The next link has no forward pagination cursor")
        except (ValueError, KeyError, TypeError):
            raise PaginationError("Starling returned an invalid next link") from None
        self.request = {**self.request, "cursor": cursor, "page_to_fetch": "NEXT"}


def iter_feed(client: StarlingClient, *, account_uid: str, category_uid: str,
              min_transaction_timestamp: datetime, max_transaction_timestamp: datetime,
              max_pages: int | None = None, request_options: RequestOptions | None = None) -> Iterator[FeedItem]:
    """Yield feed items one page at a time. Breaking the loop prevents further requests."""
    if request_options and request_options.get("additional_query_parameters"):
        raise ValueError("Pass feed filters directly; additional query parameters are unsupported by iter_feed")
    pagination = _FeedPagination(client.base_url, account_uid, category_uid, min_transaction_timestamp, max_transaction_timestamp, max_pages)
    count = 0
    while True:
        page = client.feed.list_page(**pagination.take(), request_options=request_options)
        yield from page.feed_items
        count += 1
        if not page.links.next or (max_pages is not None and count >= max_pages):
            return
        pagination.advance(page.links.next)


async def async_iter_feed(client: AsyncStarlingClient, *, account_uid: str, category_uid: str,
                          min_transaction_timestamp: datetime, max_transaction_timestamp: datetime,
                          max_pages: int | None = None, request_options: RequestOptions | None = None) -> AsyncIterator[FeedItem]:
    """Async feed iteration; cancellation stops the current request or retry wait."""
    if request_options and request_options.get("additional_query_parameters"):
        raise ValueError("Pass feed filters directly; additional query parameters are unsupported by async_iter_feed")
    pagination = _FeedPagination(client.base_url, account_uid, category_uid, min_transaction_timestamp, max_transaction_timestamp, max_pages)
    count = 0
    while True:
        page = await client.feed.list_page(**pagination.take(), request_options=request_options)
        for item in page.feed_items:
            yield item
        count += 1
        if not page.links.next or (max_pages is not None and count >= max_pages):
            return
        pagination.advance(page.links.next)
